File: android/src/main/java/com/alpha0010/pdf/TextAnnotationHandler.kt
package com.alpha0010.pdf

import android.content.Context
import android.graphics.Color
import android.graphics.PointF
import android.graphics.RectF
import android.text.InputType
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.TextView
import java.util.UUID
import kotlin.math.hypot

/**
 * Delegate interface for TextAnnotationHandler.
 * Provides view-specific operations.
 */
interface TextAnnotationHandlerDelegate {
    val textHandlerDrawingController: DrawingController
    val textHandlerContentContainer: View
    val textHandlerHostView: ViewGroup
    val textHandlerZoomScale: Float
    fun textHandlerContentRectForPage(page: Int): RectF
    fun textHandlerPageForPoint(pointInContent: PointF): Int
    fun textHandlerRedrawOverlay()
}

/**
 * Reusable handler for text annotation input and dragging.
 * Used by both ZoomablePdfScrollView and PagingPdfView.
 */
class TextAnnotationHandler(private val context: Context) {

    var delegate: TextAnnotationHandlerDelegate? = null
    var textColor: String = "#0000FF"
    var textFontSize: Float = 16f

    // Text input state
    private var textInputView: EditText? = null
    private var textInputPage: Int = 0
    private var textInputNormalizedPoint: PointF = PointF()

    // Text drag state
    var isDraggingText: Boolean = false
        private set
    private var draggingTextPage: Int = 0
    private var draggingLabel: TextView? = null
    private var draggingTouchOffsetX: Float = 0f
    private var draggingTouchOffsetY: Float = 0f
    private var draggingText: DrawingText? = null

    // Pending touch: tap vs drag detection
    private var pendingText: DrawingText? = null       // existing text hit
    private var pendingNewTextPoint: PointF? = null     // new text tap (no hit)
    private var pendingNewTextEvent: MotionEvent? = null
    private var pendingTextPage: Int = 0
    private var pendingTouchStartX: Float = 0f
    private var pendingTouchStartY: Float = 0f
    private val dragThreshold: Float = 24f  // pixels
    private var pendingWasMultiTouch = false

    // Editing existing text
    private var editingTextId: String? = null
    private var editingTextColor: String? = null
    private var editingTextFontSize: Float? = null

    val hasActiveTextInput: Boolean get() = textInputView != null
    val hasPendingText: Boolean get() = pendingText != null || pendingNewTextPoint != null

    // Convert screen touch coordinates to content coordinates
    var screenToContentConverter: ((MotionEvent) -> PointF)? = null
    // Convert content coordinates back to screen (hostView) coordinates
    var contentToScreenConverter: ((PointF) -> PointF)? = null

    private fun screenToContent(event: MotionEvent): PointF {
        return screenToContentConverter?.invoke(event) ?: PointF(event.x, event.y)
    }

    private fun contentToScreen(contentPoint: PointF): PointF {
        return contentToScreenConverter?.invoke(contentPoint) ?: contentPoint
    }

    /**
     * Handle touch in text mode. Returns true if touch was consumed.
     */
    fun handleTouchDown(event: MotionEvent): Boolean {
        val delegate = delegate ?: return false

        // Clear any stale pending state from previous gesture
        if (hasPendingText) clearPending()

        // If there's a text input view, check if touch is outside it
        val existingInput = textInputView
        if (existingInput != null) {
            val loc = IntArray(2)
            existingInput.getLocationOnScreen(loc)
            val inputRect = RectF(
                loc[0].toFloat(), loc[1].toFloat(),
                (loc[0] + existingInput.width).toFloat(),
                (loc[1] + existingInput.height).toFloat()
            )
            val screenLoc = IntArray(2)
            delegate.textHandlerHostView.getLocationOnScreen(screenLoc)
            val touchScreenX = event.x + screenLoc[0]
            val touchScreenY = event.y + screenLoc[1]

            if (!inputRect.contains(touchScreenX, touchScreenY)) {
                commitTextInput()
                return true
            }
            return false
        }

        // Convert touch to content coordinates
        val contentPoint = screenToContent(event)
        val page = delegate.textHandlerPageForPoint(contentPoint)
        val pageRect = delegate.textHandlerContentRectForPage(page)
        if (pageRect.isEmpty) return false

        // Normalize
        val normalizedX = (contentPoint.x - pageRect.left) / pageRect.width()
        val normalizedY = (contentPoint.y - pageRect.top) / pageRect.height()
        val normalizedPoint = PointF(normalizedX, normalizedY)

        // Check if touching existing text — defer to detect tap vs drag
        val hitResult = hitTestTextAnnotation(normalizedPoint, page)
        if (hitResult != null) {
            val (_, textAnnotation) = hitResult
            pendingText = textAnnotation
            pendingTextPage = page
            pendingTouchStartX = event.x
            pendingTouchStartY = event.y
            return true
        }

        // No existing text — defer until touchUp to avoid opening on multi-finger
        pendingNewTextPoint = normalizedPoint
        pendingNewTextEvent = MotionEvent.obtain(event)
        pendingTextPage = page
        pendingTouchStartX = event.x
        pendingTouchStartY = event.y
        return true
    }

    fun handleTouchMove(event: MotionEvent): Boolean {
        val delegate = delegate ?: return false

        // Pending new text — cancel if finger moves (scroll/zoom gesture)
        if (pendingNewTextPoint != null) {
            val dist = hypot(event.x - pendingTouchStartX, event.y - pendingTouchStartY)
            if (dist >= dragThreshold) {
                pendingWasMultiTouch = true
                clearPending()
            }
            return true
        }

        // Pending existing text — check if finger moved enough to start drag
        val text = pendingText
        if (text != null) {
            val dist = hypot(event.x - pendingTouchStartX, event.y - pendingTouchStartY)
            if (dist >= dragThreshold) {
                startDraggingText(text, pendingTextPage, event)
                clearPending()
            }
            return true
        }

        if (!isDraggingText) return false
        val label = draggingLabel ?: return false

        val newX = (event.x - draggingTouchOffsetX).toInt()
        val newY = (event.y - draggingTouchOffsetY).toInt()
        label.layout(newX, newY, newX + label.width, newY + label.height)
        return true
    }

    fun handleTouchUp(event: MotionEvent): Boolean {
        // Pending existing text, no drag → tap to edit
        val text = pendingText
        if (text != null) {
            editExistingText(text, pendingTextPage)
            clearPending()
            return true
        }

        // Pending new text → tap to create
        val point = pendingNewTextPoint
        val ev = pendingNewTextEvent
        if (point != null && ev != null) {
            showTextInput(point, pendingTextPage, ev)
            clearPending()
            return true
        }

        if (!isDraggingText) return false
        finishDraggingText(event)
        return true
    }

    fun handleTouchCancel() {
        // Pending existing text cancelled (single finger) → treat as tap to edit
        val text = pendingText
        if (text != null && !pendingWasMultiTouch) {
            editExistingText(text, pendingTextPage)
            clearPending()
            return
        }

        // Pending new text cancelled — only open if not multi-touch
        val point = pendingNewTextPoint
        val ev = pendingNewTextEvent
        if (point != null && ev != null && !pendingWasMultiTouch) {
            showTextInput(point, pendingTextPage, ev)
            clearPending()
            return
        }

        if (hasPendingText) {
            clearPending()
            return
        }

        if (!isDraggingText) return
        // Restore text at original position
        val dragText = draggingText
        val del = delegate
        if (dragText != null && del != null) {
            del.textHandlerDrawingController.addText(dragText, draggingTextPage)
            del.textHandlerRedrawOverlay()
        }
        cancelDraggingText()
    }

    fun handleMultiTouchDetected() {
        pendingWasMultiTouch = true
        clearPending()
    }

    private fun clearPending() {
        pendingText = null
        pendingNewTextEvent?.recycle()
        pendingNewTextEvent = null
        pendingNewTextPoint = null
        pendingWasMultiTouch = false
    }

    // MARK: - Edit Existing Text

    private fun editExistingText(text: DrawingText, page: Int) {
        val delegate = delegate ?: return

        // Remove text so it doesn't render while editing
        delegate.textHandlerDrawingController.removeText(text.id, page)
        delegate.textHandlerRedrawOverlay()

        val normalizedPoint = PointF(text.point[0], text.point[1])
        editingTextId = text.id
        editingTextColor = text.color
        editingTextFontSize = text.fontSize

        // Convert text position to screen coordinates for the EditText
        val pageRect = delegate.textHandlerContentRectForPage(page)
        val contentX = pageRect.left + text.point[0] * pageRect.width()
        val contentY = pageRect.top + text.point[1] * pageRect.height()
        val screenPos = contentToScreen(PointF(contentX, contentY))
        val fakeEvent = MotionEvent.obtain(0, 0, MotionEvent.ACTION_DOWN, screenPos.x, screenPos.y, 0)

        showTextInput(normalizedPoint, page, fakeEvent, text.str, text.color, text.fontSize)
        fakeEvent.recycle()
    }

    // MARK: - Text Input

    private fun showTextInput(normalizedPoint: PointF, page: Int, event: MotionEvent,
                              existingText: String? = null, color: String? = null, fontSize: Float? = null) {
        val delegate = delegate ?: return

        textInputPage = page
        textInputNormalizedPoint = normalizedPoint

        val hostView = delegate.textHandlerHostView
        val zoomScale = delegate.textHandlerZoomScale
        val useFontSize = fontSize ?: textFontSize
        val useColor = color ?: textColor

        val screenX = event.x
        val screenY = event.y

        val editText = EditText(context).apply {
            setBackgroundColor(Color.TRANSPARENT)
            textSize = useFontSize * zoomScale
            setTextColor(try { Color.parseColor(useColor) } catch (e: Exception) { Color.BLUE })
            setPadding(0, 0, 0, 0)
            gravity = Gravity.TOP or Gravity.START
            // Multi-line, no auto-suggestions, no word-wrap
            inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_FLAG_MULTI_LINE or InputType.TYPE_TEXT_FLAG_NO_SUGGESTIONS
            // Allow horizontal scrolling (no word-wrap)
            setHorizontallyScrolling(true)
            maxLines = Int.MAX_VALUE
            isSingleLine = false
            if (existingText != null) {
                setText(existingText)
                setSelection(existingText.length)
            }
        }

        val maxWidthScreen = (hostView.width - screenX.toInt()).coerceAtLeast(120)
        val lp = FrameLayout.LayoutParams(maxWidthScreen, ViewGroup.LayoutParams.WRAP_CONTENT)
        lp.leftMargin = screenX.toInt()
        lp.topMargin = screenY.toInt()

        hostView.addView(editText, lp)
        editText.measure(
            View.MeasureSpec.makeMeasureSpec(maxWidthScreen, View.MeasureSpec.EXACTLY),
            View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
        )
        editText.layout(lp.leftMargin, lp.topMargin, lp.leftMargin + editText.measuredWidth, lp.topMargin + editText.measuredHeight)
        editText.bringToFront()
        editText.requestFocus()

        // Re-measure on text change so height grows with newlines
        editText.addTextChangedListener(object : android.text.TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: android.text.Editable?) {
                editText.post {
                    val lineHeight = editText.lineHeight
                    val lineCount = editText.lineCount.coerceAtLeast(1)
                    val newHeight = lineCount * lineHeight + editText.paddingTop + editText.paddingBottom
                    editText.layout(lp.leftMargin, lp.topMargin, lp.leftMargin + maxWidthScreen, lp.topMargin + newHeight)
                }
            }
        })

        // Show keyboard
        val imm = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        editText.postDelayed({ imm.showSoftInput(editText, InputMethodManager.SHOW_IMPLICIT) }, 100)

        textInputView = editText
    }

    fun commitTextInput() {
        val editText = textInputView ?: return
        val delegate = delegate ?: return

        val text = editText.text?.toString()?.trim() ?: ""

        // Hide keyboard
        val imm = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(editText.windowToken, 0)

        (editText.parent as? ViewGroup)?.removeView(editText)
        textInputView = null

        val existingId = editingTextId
        val existingColor = editingTextColor ?: textColor
        val existingFontSize = editingTextFontSize
        editingTextId = null
        editingTextColor = null
        editingTextFontSize = null

        if (text.isEmpty()) return

        val useFontSize = existingFontSize ?: textFontSize
        val useColor = if (existingId != null) existingColor else textColor
        val drawingText = DrawingText(
            id = existingId ?: UUID.randomUUID().toString(),
            color = useColor,
            fontSize = useFontSize,
            point = listOf(textInputNormalizedPoint.x, textInputNormalizedPoint.y),
            str = text
        )
        delegate.textHandlerDrawingController.addTextWithUndo(drawingText, textInputPage)
        delegate.textHandlerRedrawOverlay()
    }

    // MARK: - Hit Test

    private fun hitTestTextAnnotation(normalizedPoint: PointF, page: Int): Pair<Int, DrawingText>? {
        val delegate = delegate ?: return null

        val texts = delegate.textHandlerDrawingController.getTexts(page)
        val pageRect = delegate.textHandlerContentRectForPage(page)
        if (pageRect.isEmpty) return null

        for ((index, text) in texts.withIndex()) {
            if (text.point.size < 2) continue

            val textX = text.point[0]
            val textY = text.point[1]

            val paint = android.text.TextPaint().apply {
                textSize = text.fontSize * context.resources.displayMetrics.density
                isAntiAlias = true
            }
            val layout = android.text.StaticLayout.Builder.obtain(text.str, 0, text.str.length, paint, 100000)
                .setAlignment(android.text.Layout.Alignment.ALIGN_NORMAL)
                .setIncludePad(false)
                .build()
            var maxLineWidth = 0f
            for (i in 0 until layout.lineCount) {
                maxLineWidth = maxOf(maxLineWidth, layout.getLineWidth(i))
            }
            val normalizedWidth = maxLineWidth / pageRect.width()
            val normalizedHeight = layout.height.toFloat() / pageRect.height()

            val padding = 0.02f
            val hitRect = RectF(
                textX - padding,
                textY - padding,
                textX + normalizedWidth + padding * 2,
                textY + normalizedHeight + padding * 2
            )

            if (hitRect.contains(normalizedPoint.x, normalizedPoint.y)) {
                return Pair(index, text)
            }
        }
        return null
    }

    // MARK: - Text Dragging

    private fun startDraggingText(textAnnotation: DrawingText, page: Int, event: MotionEvent) {
        val delegate = delegate ?: return

        isDraggingText = true
        draggingTextPage = page
        draggingText = textAnnotation

        // Remove text from controller so overlay doesn't draw it
        delegate.textHandlerDrawingController.removeText(textAnnotation.id, page)
        delegate.textHandlerRedrawOverlay()

        val zoomScale = delegate.textHandlerZoomScale
        val pageRect = delegate.textHandlerContentRectForPage(page)
        val hostView = delegate.textHandlerHostView

        // Create floating label (multi-line)
        val label = TextView(context).apply {
            text = textAnnotation.str
            textSize = textAnnotation.fontSize * zoomScale
            setTextColor(
                (try { Color.parseColor(textAnnotation.color) } catch (e: Exception) { Color.BLUE })
                    .let { Color.argb(180, Color.red(it), Color.green(it), Color.blue(it)) }
            )
            setBackgroundColor(Color.TRANSPARENT)
        }

        // Convert text position from content to screen coordinates
        val contentX = pageRect.left + textAnnotation.point[0] * pageRect.width()
        val contentY = pageRect.top + textAnnotation.point[1] * pageRect.height()
        val screenPos = contentToScreen(PointF(contentX, contentY))
        val screenX = screenPos.x
        val screenY = screenPos.y

        val widthSpec = View.MeasureSpec.makeMeasureSpec(hostView.width, View.MeasureSpec.AT_MOST)
        val heightSpec = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
        label.measure(widthSpec, heightSpec)
        val lw = label.measuredWidth
        val lh = label.measuredHeight
        val left = screenX.toInt()
        val top = screenY.toInt()

        val lp = FrameLayout.LayoutParams(lw, lh)
        hostView.addView(label, lp)
        label.layout(left, top, left + lw, top + lh)
        label.bringToFront()

        draggingTouchOffsetX = event.x - screenX
        draggingTouchOffsetY = event.y - screenY
        draggingLabel = label
    }

    private fun finishDraggingText(event: MotionEvent) {
        val text = draggingText ?: return
        val delegate = delegate ?: run { cancelDraggingText(); return }

        val labelX = event.x - draggingTouchOffsetX
        val labelY = event.y - draggingTouchOffsetY

        val contentPoint = screenToContentConverter?.invoke(
            MotionEvent.obtain(0, 0, MotionEvent.ACTION_DOWN, labelX, labelY, 0)
        ) ?: PointF(labelX, labelY)

        val pageRect = delegate.textHandlerContentRectForPage(draggingTextPage)
        val normalizedX = (contentPoint.x - pageRect.left) / pageRect.width()
        val normalizedY = (contentPoint.y - pageRect.top) / pageRect.height()

        val newPoint = listOf(normalizedX, normalizedY)
        // Re-add text at original position (it was removed during drag start)
        delegate.textHandlerDrawingController.addText(text, draggingTextPage)
        // Record the move as a single undoable action
        delegate.textHandlerDrawingController.moveTextWithUndo(text.id, text.point, newPoint, draggingTextPage)
        cancelDraggingText()
        delegate.textHandlerRedrawOverlay()
    }

    private fun cancelDraggingText() {
        draggingLabel?.let { (it.parent as? ViewGroup)?.removeView(it) }
        draggingLabel = null
        isDraggingText = false
        draggingText = null
    }
}

File: ios/TextAnnotationHandler.swift
import UIKit

// MARK: - Protocol

protocol TextAnnotationHandlerDelegate: AnyObject {
    var textHandlerDrawingController: DrawingController { get }
    var textHandlerContentContainer: UIView { get }
    /// View for adding floating drag label
    var textHandlerHostView: UIView { get }
    var textHandlerZoomScale: CGFloat { get }

    func textHandlerContentRectForPage(_ page: Int) -> CGRect
    func textHandlerPageForPoint(_ pointInContent: CGPoint) -> Int
    func textHandlerRedrawOverlay()
}

// MARK: - TextAnnotationHandler

class TextAnnotationHandler: NSObject, UITextViewDelegate {

    weak var delegate: TextAnnotationHandlerDelegate?

    var textColor: String = "#0000FF"
    var textFontSize: CGFloat = 16.0

    // MARK: - Text Input State

    private var textInputView: UITextView?
    private var textInputPage: Int = 0
    private var textInputNormalizedPoint: CGPoint = .zero

    // MARK: - Text Drag State

    private(set) var isDraggingText: Bool = false
    private var draggingTextPage: Int = 0
    private var draggingLabel: UILabel?
    private var draggingTouchOffset: CGPoint = .zero
    private var draggingText: DrawingText?

    // Pending touch: tap vs multi-finger detection
    private var pendingText: DrawingText? // existing text hit
    private var pendingNewTextPoint: CGPoint? // new text tap (no hit)
    private var pendingTextPage: Int = 0
    private var pendingTouchStart: CGPoint = .zero
    private let dragThreshold: CGFloat = 8.0
    private var pendingWasMultiTouch = false

    // Editing existing text
    private var editingTextId: String?
    private var editingTextColor: String?

    var hasActiveTextInput: Bool {
        return textInputView != nil
    }

    var hasPendingText: Bool {
        return pendingText != nil || pendingNewTextPoint != nil
    }

    // MARK: - Touch Handling

    /// Handle touch began in text mode. Returns true if touch was consumed.
    func handleTouchBegan(_ touch: UITouch) -> Bool {
        guard let delegate = delegate else { return false }

        let contentContainer = delegate.textHandlerContentContainer

        // If there's a text input view, check if touch is outside it
        if let textView = textInputView {
            let touchInContent = touch.location(in: contentContainer)
            if !textView.frame.contains(touchInContent) {
                commitTextInput()
                return true // consume this touch (just committed)
            }
            return false // touch inside text view, let it handle
        }

        // Clear any stale pending state from previous gesture
        if hasPendingText { clearPending() }

        let location = touch.location(in: contentContainer)
        let page = delegate.textHandlerPageForPoint(location)
        let pageRect = delegate.textHandlerContentRectForPage(page)
        guard !pageRect.isEmpty else { return false }

        // Normalize to 0-1
        let normalizedX = location.x / pageRect.width
        let normalizedY = (location.y - pageRect.minY) / pageRect.height
        let normalizedPoint = CGPoint(x: normalizedX, y: normalizedY)

        // Check if touching an existing text annotation
        if let (textIndex, textAnnotation) = hitTestTextAnnotation(at: normalizedPoint, page: page) {
            // Don't start drag immediately — wait for movement to distinguish tap vs drag
            pendingText = textAnnotation
            pendingTextPage = page
            pendingTouchStart = touch.location(in: delegate.textHandlerHostView)
            return true
        }

        // No existing text — defer until touchEnded to avoid opening on multi-finger
        pendingNewTextPoint = normalizedPoint
        pendingTextPage = page
        pendingTouchStart = touch.location(in: delegate.textHandlerHostView)
        return true
    }

    func handleTouchMoved(_ touch: UITouch) {
        guard let delegate = delegate else { return }

        // If pending new text and finger moves, mark as moved (not a clean tap)
        if pendingNewTextPoint != nil {
            let location = touch.location(in: delegate.textHandlerHostView)
            let dist = hypot(location.x - pendingTouchStart.x, location.y - pendingTouchStart.y)
            if dist >= dragThreshold {
                pendingWasMultiTouch = true // finger moved = not a tap
                clearPending()
            }
            return
        }

        // If we have a pending text hit, check if finger moved enough to start drag
        if let text = pendingText {
            let location = touch.location(in: delegate.textHandlerHostView)
            let dist = hypot(location.x - pendingTouchStart.x, location.y - pendingTouchStart.y)
            if dist >= dragThreshold {
                startDraggingText(
                    textAnnotation: text,
                    textIndex: 0,
                    page: pendingTextPage,
                    touch: touch
                )
                clearPending()
            }
            return
        }

        guard isDraggingText else { return }
        let location = touch.location(in: delegate.textHandlerHostView)
        guard let label = draggingLabel else { return }
        label.frame.origin = CGPoint(
            x: location.x - draggingTouchOffset.x,
            y: location.y - draggingTouchOffset.y
        )
    }

    func handleTouchEnded(_ touch: UITouch) {
        // If pending existing text and no drag started — this is a tap → edit
        if let text = pendingText {
            editExistingText(text, page: pendingTextPage)
            clearPending()
            return
        }

        // If pending new text — this is a tap → show input
        if let point = pendingNewTextPoint {
            showTextInput(at: point, page: pendingTextPage)
            clearPending()
            return
        }

        guard isDraggingText, let _ = delegate else { return }
        finishDraggingText(touch: touch)
    }

    func handleMultiTouchDetected() {
        // Called by callers when 2+ fingers detected — ensures pending new text won't open
        pendingWasMultiTouch = true
        clearPending()
    }

    func handleTouchCancelled() {
        // If pending existing text (finger didn't move enough to drag), treat cancel as tap → edit
        if let text = pendingText {
            editExistingText(text, page: pendingTextPage)
            clearPending()
            return
        }
        // If pending new text: open input only if it wasn't a multi-touch/move gesture
        if let point = pendingNewTextPoint {
            if pendingWasMultiTouch {
                // multi-touch: don't open text input
            } else {
                showTextInput(at: point, page: pendingTextPage)
            }
            clearPending()
            return
        }
        guard isDraggingText else { return }
        // Restore text at original position on cancel
        if let text = draggingText, let delegate = delegate {
            delegate.textHandlerDrawingController.addText(text, toPage: draggingTextPage)
            delegate.textHandlerRedrawOverlay()
        }
        cancelDraggingText()
    }

    private func clearPending() {
        pendingText = nil
        pendingNewTextPoint = nil
        pendingWasMultiTouch = false
    }

    // MARK: - Edit Existing Text

    private func editExistingText(_ text: DrawingText, page: Int) {
        guard let delegate = delegate else { return }

        // Remove text from model so it doesn't render while editing
        delegate.textHandlerDrawingController.removeText(withId: text.id, onPage: page)
        delegate.textHandlerRedrawOverlay()

        // Show input at the text's position with its content
        let normalizedPoint = CGPoint(x: CGFloat(text.point[0]), y: CGFloat(text.point[1]))
        editingTextId = text.id
        editingTextColor = text.color
        showTextInput(at: normalizedPoint, page: page, existingText: text.str, color: text.color, fontSize: text.fontSize)
    }

    // MARK: - Text Input

    private func showTextInput(at normalizedPoint: CGPoint, page: Int, existingText: String? = nil, color: String? = nil, fontSize: CGFloat? = nil) {
        guard let delegate = delegate else { return }

        textInputPage = page
        textInputNormalizedPoint = normalizedPoint

        let pageRect = delegate.textHandlerContentRectForPage(page)
        let contentContainer = delegate.textHandlerContentContainer

        // Position in content container coordinates (unzoomed)
        let contentX = normalizedPoint.x * pageRect.width
        let contentY = pageRect.minY + normalizedPoint.y * pageRect.height
        let maxWidthInContent = pageRect.width - contentX

        let useFontSize = fontSize ?? textFontSize
        let useColor = color ?? textColor

        let textView = UITextView()
        textView.backgroundColor = .clear
        textView.font = UIFont.systemFont(ofSize: useFontSize)
        textView.textColor = UIColor(hexString: useColor) ?? .blue
        if let existingText = existingText {
            textView.text = existingText
        }
        textView.isScrollEnabled = true
        textView.textContainerInset = .zero
        textView.textContainer.lineFragmentPadding = 0
        textView.textContainer.widthTracksTextView = false
        textView.textContainer.size = CGSize(width: 100000, height: 100000)
        textView.returnKeyType = .default
        textView.delegate = self

        let textViewWidth = max(60, maxWidthInContent)
        textView.frame = CGRect(x: contentX, y: contentY, width: textViewWidth, height: useFontSize + 8)
        textView.autoresizingMask = []

        contentContainer.addSubview(textView)

        // Resize to fit existing content (for multi-line text editing)
        if existingText != nil {
            let fitSize = textView.sizeThatFits(CGSize(width: textView.frame.width, height: CGFloat.greatestFiniteMagnitude))
            textView.frame.size.height = max(fitSize.height, useFontSize + 8)
        }

        textView.becomeFirstResponder()

        textInputView = textView

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(textInputDidChange(_:)),
            name: UITextView.textDidChangeNotification,
            object: textView
        )
    }

    @objc private func textInputDidChange(_ notification: Notification) {
        guard let textView = notification.object as? UITextView,
              textView === textInputView else { return }

        let newSize = textView.sizeThatFits(CGSize(width: textView.frame.width, height: CGFloat.greatestFiniteMagnitude))
        textView.frame.size.height = max(newSize.height, (textView.font?.pointSize ?? textFontSize) + 8)
    }

    func commitTextInput() {
        guard let textView = textInputView, let delegate = delegate else { return }

        NotificationCenter.default.removeObserver(self, name: UITextView.textDidChangeNotification, object: textView)

        let text = textView.text?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let inputFontSize = textView.font?.pointSize ?? textFontSize

        textView.resignFirstResponder()
        textView.removeFromSuperview()
        textInputView = nil

        let existingId = editingTextId
        let existingColor = editingTextColor ?? textColor
        editingTextId = nil
        editingTextColor = nil

        guard !text.isEmpty else { return }

        let drawingText = DrawingText(
            id: existingId ?? UUID().uuidString,
            color: existingId != nil ? existingColor : textColor,
            fontSize: inputFontSize,
            point: [textInputNormalizedPoint.x, textInputNormalizedPoint.y],
            str: text
        )
        delegate.textHandlerDrawingController.addTextWithUndo(drawingText, toPage: textInputPage)
        delegate.textHandlerRedrawOverlay()
    }

    // MARK: - UITextViewDelegate

    // MARK: - Hit Test

    private func hitTestTextAnnotation(at normalizedPoint: CGPoint, page: Int) -> (Int, DrawingText)? {
        guard let delegate = delegate else { return nil }

        let texts = delegate.textHandlerDrawingController.getTexts(forPage: page)
        let pageRect = delegate.textHandlerContentRectForPage(page)
        guard !pageRect.isEmpty else { return nil }

        for (index, text) in texts.enumerated() {
            guard text.point.count >= 2 else { continue }

            let textX = text.point[0]
            let textY = text.point[1]

            let attributes: [NSAttributedString.Key: Any] = [
                .font: UIFont.systemFont(ofSize: text.fontSize)
            ]
            let attrStr = NSAttributedString(string: text.str, attributes: attributes)
            let textSize = attrStr.boundingRect(
                with: CGSize(width: 100000, height: 100000),
                options: [.usesLineFragmentOrigin],
                context: nil
            ).size

            let normalizedWidth = textSize.width / pageRect.width
            let normalizedHeight = textSize.height / pageRect.height

            let padding: CGFloat = 0.02
            let hitRect = CGRect(
                x: textX - padding,
                y: textY - padding,
                width: normalizedWidth + padding * 2,
                height: normalizedHeight + padding * 2
            )

            if hitRect.contains(normalizedPoint) {
                return (index, text)
            }
        }
        return nil
    }

    // MARK: - Text Dragging

    private func startDraggingText(textAnnotation: DrawingText, textIndex: Int, page: Int, touch: UITouch) {
        guard let delegate = delegate else { return }

        isDraggingText = true
        draggingTextPage = page

        // Remove text from controller so overlay doesn't draw it
        draggingText = textAnnotation
        delegate.textHandlerDrawingController.removeText(withId: textAnnotation.id, onPage: page)
        delegate.textHandlerRedrawOverlay()

        let zoomScale = delegate.textHandlerZoomScale
        let pageRect = delegate.textHandlerContentRectForPage(page)
        let hostView = delegate.textHandlerHostView
        let touchInHost = touch.location(in: hostView)

        // Create floating label
        let label = UILabel()
        label.text = textAnnotation.str
        label.font = UIFont.systemFont(ofSize: textAnnotation.fontSize * zoomScale)
        label.textColor = (UIColor(hexString: textAnnotation.color) ?? .blue).withAlphaComponent(0.7)
        label.numberOfLines = 0
        label.lineBreakMode = .byClipping
        label.backgroundColor = .clear
        label.sizeToFit()

        // Convert text position from content to host view coordinates
        let contentContainer = delegate.textHandlerContentContainer
        let contentPt = CGPoint(
            x: textAnnotation.point[0] * pageRect.width,
            y: pageRect.minY + textAnnotation.point[1] * pageRect.height
        )
        let screenPt = contentContainer.convert(contentPt, to: hostView)
        label.frame.origin = screenPt

        // Store offset between touch and label origin in host view coordinates (for label movement)
        draggingTouchOffset = CGPoint(x: touchInHost.x - screenPt.x, y: touchInHost.y - screenPt.y)

        hostView.addSubview(label)
        draggingLabel = label
    }

    private func finishDraggingText(touch: UITouch) {
        guard isDraggingText, let text = draggingText, let delegate = delegate else {
            cancelDraggingText()
            return
        }

        // Use touch location in content container directly — avoids contentOffset dependency
        let contentContainer = delegate.textHandlerContentContainer
        let locationInContent = touch.location(in: contentContainer)

        let pageRect = delegate.textHandlerContentRectForPage(draggingTextPage)

        // Apply the touch offset in content coordinates
        let zoomScale = delegate.textHandlerZoomScale
        let contentOffsetX = draggingTouchOffset.x / zoomScale
        let contentOffsetY = draggingTouchOffset.y / zoomScale

        let contentX = locationInContent.x - contentOffsetX
        let contentY = locationInContent.y - contentOffsetY

        let normalizedX = contentX / pageRect.width
        let normalizedY = (contentY - pageRect.minY) / pageRect.height

        let newPoint: [CGFloat] = [normalizedX, normalizedY]
        // Re-add the text at original position first (it was removed during drag start)
        delegate.textHandlerDrawingController.addText(text, toPage: draggingTextPage)
        // Then record the move as a single undoable action
        delegate.textHandlerDrawingController.moveTextWithUndo(withId: text.id, fromPoint: text.point, toPoint: newPoint, onPage: draggingTextPage)
        cancelDraggingText()
        delegate.textHandlerRedrawOverlay()
    }

    private func cancelDraggingText() {
        draggingLabel?.removeFromSuperview()
        draggingLabel = nil
        isDraggingText = false
        draggingText = nil
    }
}
